### PR TITLE
Deep git integration

### LINKED ISSUE #43252: A list of things spack can do better
### Summary

In https://github.com/NixOS/nix/issues/10253, I said that "Spack is like nix except significantly worse in every conceivable way." This was, admittedly, far too harsh, but there's a few UX issues that desperately need fixing:

## Spack runs when the `.git` folder is missing

The `.git` folder is essential for spack to be able to update.The cluser-provided spack did not copy the `.git` folder over (and I think, still doesn't), and would also be needed for a few of my other suggestions.

Spack should check for a `.git` folder, and if it's missing, refuse to run until it's been populated. Alternatively, it should try to autodetect its tag and replace the `.git` folder for the user.

## Upgrading spack is not the obvious way to get more up-to-date packages. 

My lab was using the cluster-provided spack `0.18.1` for the longest time (i.e. until 4 months ago), and didn't know that updating spack was the way to get access to more recent versions of packages.

I suggest that spack should automatically `git fetch` every when it's invoked (at most once every 24h). spack should check if there's a new tag available, and if there is, `spack` should issue a warning every time it's invoked telling users to upgrade.

Additionally, I feel like the package repository should be its own thing and update independently of the package manager. This git repository is huge.

## Upgrading spack is 'hard'

Even ignoring the fact that I had to do some creative work to reconstruct the `.git` folder for the original spack installation, the spack upgrade method isn't intuitive to those that aren't familliar with git. I ended up creating a step-by-step guide to explain the process of finding the latest release tag and running `git checkout <tag>`. Additionally, due to our lab's policy, I have to copy over the entire 91 gigabytes of spack folder into a new folder, and *then* `git checkout <tag>` in the new folder. This takes about 8 hours.

There's already an issue and PR open for this, although the PR has stagnated it seems: https://github.com/spack/spack/issues/9134, https://github.com/spack/spack/pull/37549

### Rationale

_No response_

### Description

_No response_

### Additional information

If it would help, I would be willing to write some of this code.

### General information

- [X] I have searched the issues of this repo and believe this is not a duplicate

### PR DESCRIPTION
**Pull request by Pandapip1**
Fixes https://github.com/spack/spack/issues/43252 (automatic upgrades are out of scope).

### CODE REVIEW

**Comment by Pandapip1** [comment]
@spackbot fix style

**Comment by Pandapip1** [comment]
@spackbot rerun

File: lib/spack/spack/main.py [hunk 1]
```diff
@@ -932,10 +1120,16 @@ def _main(argv=None):
...
-    # configuration. This doesn't include much -- setting up th parser,
+    # configuration. This doesn't include much -- setting up the parser,
     # restoring some key environment variables, very simple CLI options, etc.
     # ------------------------------------------------------------------------
 
+    # We want the git warnings to ALWAYS be shown before anything else can happen
+    status = perform_git_checks()
+    if status != 0:
+        return status
+    del status
```
Comment (alalazo) [inline lib/spack/spack/main.py:1131]: I don't think we want this (require to have `git` available and potentially wait a long time on each command).

**Review by alalazo — CHANGES_REQUESTED** [review]

**Comment by alalazo** [comment]
Closing, since we don't want to require `git` for using Spack. Also, reading #43252 it seems most of the issues were due to not being aware of a few features (e.g. customization of the install tree etc.)

**Pull request closed without merging**